src/traverse/index.js

```diff
@@ -193,22 +199,82 @@ module.exports = {
           }
 
           // Per-node handler.
-          if (typeof handler[node.type] === 'function') {
+          let handlerFunc_pre = null;
+          if (handler[node.type] &&
+              typeof handler[node.type] === 'function') {
+            handlerFunc_pre = handler[node.type].bind(handler);
+          }
+          else if (handler[node.type] &&
+                   typeof handler[node.type] === 'object' &&
+                   typeof handler[node.type].pre === 'function') {
+            handlerFunc_pre = handler[node.type].pre.bind(handler);
+          }
+
+          if (handlerFunc_pre) {
             if (options.asNodes) {
-              handler[node.type](node, parent, prop, index);
+              handlerFunc_pre(node, parent, prop, index);
             } else {
               // A path/node can be removed by some previous handler.
               if (!nodePath.isRemoved()) {
-                const handlerResult = handler[node.type](nodePath);
+                const handlerResult = handlerFunc_pre(nodePath);
                 // Explicitly stop traversal.
                 if (handlerResult === false) {
                   return false;
                 }
               }
             }
           }
+        } // Loop over handlers
+
+      }, // pre func
+
+      /**
+       * Handler on node exit.
+       */
+      post(node, parent, prop, index) {
+        if (!node) { // No node? Not sure how this can happen, but without it many tests fail.
```

Comment:
Hm.. interesting. I can't recall specific issue from top of my head why the `node` can be `null` here -- I guess in cases when it's got removed maybe. But yeah, we can leave this `if` here, and remove the mystic comment.